### ISSUE TITLE
Feature Request: Export current module from AVA

### ISSUE THREAD

**Issue by lukechilds** [avajs/ava#1460]
It gets kind of annoying having to import the module you want to test with:

```js
import module from '../';
```

Especially if you move a test into a sub folder and then have to change it to:

```js
import module from '../../';
```

I think it would be really neat if AVA could expose the current module itself, so:

```js
import test from 'ava';
import module from '../../';
```

Would become:

```js
import { test, module } from 'ava';
```

Happy to put together a PR if this is something you're interested in.

**Comment by 120687** [comment]
Id line me :rindupoma
Send me contact account protect bot .please..

Pada tanggal 18 Jul 2017 18:55, "Luke Childs" <notifications@github.com>
menulis:

**Comment by ORESoftware** [comment]
Interesting

**Comment by lukechilds** [comment]
Just put this together to demo the functionality: https://github.com/lukechilds/get-root-module

No docs or anything yet but pretty obvious what it's doing from the source.

I'll submit a PR to demo the functionality in AVA.

**Comment by ORESoftware** [comment]
@lukechilds I am looking at the source for get root module:

https://github.com/lukechilds/get-root-module/blob/master/src/index.js

not really sure what it's doing 

```js
const findRoot = require('find-root');

try {
	const modulePath = findRoot(process.cwd());
	module.exports = require(modulePath);
} catch (err) {
	module.exports = undefined;
}
```

I looked at the find-root project, and it's just looking for the project root directory, so you are looking to find the "index" file of a project?

**Comment by lukechilds** [comment]
@ORESoftware I explained in more detail in the PR #1461

> Traverses up the directory tree from process.cwd() (whatever dir ava was executed from on the cli). When a dir with a package.json is found, that dir is required, if no package.json is found module will be undefined.

**Comment by ORESoftware** [comment]
@lukechilds ok bro so that's cool

one alternative this, is linking your project to itself with npm link:

```bash
cd <your-project-root> && npm link
npm link your-project
```

that way you can require('your-project') from anywhere, including your test scripts

**Comment by ORESoftware** [comment]
@lukechilds there is actually a small problem with the PR

in NPM land, requiring a directory will first look at

`{package.json}.main`

if that entry is not there, it will look for `index.js`

if that's not there, no module can be returned/required, so you'd have to handle the case where nothing can be loaded

**Comment by novemberborn** [comment]
This would indeed be useful, though users could use `get-root-module` directly, right?

I'm wary of building this in to AVA, especially with the shake up that's going to occur in the next year when ES modules land in Node.js.

**Comment by lukechilds** [comment]
@ORESoftware 

> one alternative, is linking your project to itself with npm link

That works well on your personal installation but isn't portable. Tests will break on CI and on others machines if they `git clone && npm install && npm test`.

> no module can be returned/required, so you'd have to handle the case where nothing can be loaded

That's what the try/catch does.

@novemberborn 

Yeah, no reason why users can't install `get-root-module` directly. I just thought the `import { test, module } from 'ava';` syntax was really neat, but you raise a valid point about ES modules.

I think this would be a nice addition but I'm happy using `get-root-module` directly if you're worried about ES module compatibility.

**Comment by ORESoftware** [comment]
@lukechilds 

`git clone && npm install && npm link && npm link itself && npm test`

:)

**Comment by novemberborn** [comment]
> Yeah, no reason why users can't install get-root-module directly. I just thought the import { test, module } from 'ava'; syntax was really neat, but you raise a valid point about ES modules.
> 
> I think this would be a nice addition but I'm happy using get-root-module directly if you're worried about ES module compatibility.

Would love to see a guide with solutions for this problem. `get-root-module` is a good one. A more involved one is to rewrite module paths using Babel (but I'm not expecting you to do a write-up of that approach 😉 ).

**Comment by lukechilds** [comment]
Heheh, about to head to a festival for the weekend put I can put a simple recipe together when I'm back 👍

**Comment by novemberborn** [comment]
@lukechilds great, have fun!

**Comment by ORESoftware** [comment]
Seems like "project" would be a better name than "module"

**Comment by lukechilds** [comment]
Ok so I've cleaned this up and added docs.

https://github.com/lukechilds/get-root-module

I also just came across a use case where I needed to require the current module by name to test a dynamic `require` and came up with a pure JavaScript solution without using npm link.

https://github.com/lukechilds/requireable

Works nicely with AVA too: https://github.com/lukechilds/requireable#ava-usage

So currently these are both documented in their own repos and should work standalone with AVA or any other test runner. Happy to also submit a PR to AVA with a recipe for how to use both of these solutions with AVA examples if you think that would be helpful.

Also, I still think `get-root-module` would be nice to have exported from AVA. The ES module argument is valid but I personally think it's not a deal breaker because:

- It's not gonna happen instantly, it should be well documented over time and so we'll have plenty of time to implement a fix
- ES modules will break a lot of stuff so there will be lots of other people fixing the same problem, we may well be able to drop another fix straight in
- I'm committed to maintaining `get-root-module` so happy to update that with `import` support if that's all that's needed. Or submit a PR to AVA if we need to update something on this end too

I just really enjoy how neat and succinct everything is with AVA and I feel like this really fits in well with the way AVA works. Every single test file requires importing at least `test` and the module you're testing, and often nothing else. Being able to do this with a single import seems super clean to me.

**Note: These packages are gonna be renamed soon**

`get-root-module` is a pretty verbose name, gonna rename to `this` so you can `import foo from 'this';`. @rickyclegg said he's happy for me to use it, just waiting on the transfer.

Also just realised `requireable` is a typo (I'm dyslexic 😑🔫) `requirable` seems abandoned, hopefully I can get that.

**Comment by novemberborn** [comment]
@lukechilds these seem really useful! A recipe would be great, yes.

Regarding @sindresorhus' [comment](https://github.com/avajs/ava/pull/1461#issuecomment-318203116):

> AVA already knows the root, so we could just expose that instead. (See projectDir [here](https://github.com/avajs/ava/blob/31b1380194cec664b8816bb31e333cba64ff27d9/lib/cli.js#L143)) 

Is that useful for your libraries @lukechilds?

**Comment by sindresorhus** [comment]
I'm still undecided on this. I think the best path forward is having the recipe, then see how if it's used in the wild and how, and iterate there, and then consider including it in AVA itself.

**Comment by lukechilds** [comment]
@novemberborn

> > AVA already knows the root, so we could just expose that instead. (See projectDir [here](https://github.com/avajs/ava/blob/31b1380194cec664b8816bb31e333cba64ff27d9/lib/cli.js#L143))
>
> Is that useful for your libraries @lukechilds?

Not really because currently they work completely independently from AVA, so they can be used with any other test runner. Could be an option if we decide to integrate it into AVA in the future though.

 @sindresorhus Cool, I'll put a recipe together and see how it goes from there.

**Comment by novemberborn** [comment]
@lukechilds are you still interested in writing a recipe for this?

**Comment by lukechilds** [comment]
@novemberborn I am indeed, it's on my todo list.

Totally swamped with paid work atm, deadline is next Thursday so by then I'll either be dead from exhaustion or back on OSS full time, one of the two 🙃

**Comment by novemberborn** [comment]
@lukechilds I hope it'll be the second! 🤞
